permissioned-token/program/src/instruction.rs

```diff
@@ -0,0 +1,215 @@
+    #[account(5, name = "token_program", desc = "Token program")]
+    MintTo { amount: u64 },
+
+    #[account(0, writable, name = "mint")]
+    #[account(1, writable, name = "account")]
+    #[account(2, signer, name = "owner")]
+    #[account(3, signer, name = "upstream_authority")]
+    #[account(4, signer, name = "freeze_authority")]
+    #[account(5, name = "token_program", desc = "Token program")]
+    Burn { amount: u64 },
+
+    #[account(0, writable, name = "account")]
+    #[account(1, writable, name = "destination")]
+    #[account(2, name = "mint")]
+    #[account(3, signer, name = "owner")]
+    #[account(4, signer, name = "upstream_authority")]
+    #[account(5, signer, name = "freeze_authority")]
+    #[account(6, name = "token_program", desc = "Token program")]
+    CloseAccount,
+}
+
+pub fn create_initialize_mint_instruction(
+    mint: &Pubkey,
+    payer: &Pubkey,
+    upstream_authority: &Pubkey,
+    decimals: u8,
+) -> Result<Instruction, ProgramError> {
+    Ok(Instruction {
+        program_id: crate::id(),
+        accounts: vec![
+            AccountMeta::new(*mint, true),
+            AccountMeta::new(*payer, true),
+            AccountMeta::new_readonly(*upstream_authority, false),
+            AccountMeta::new_readonly(system_program::id(), false),
+            AccountMeta::new_readonly(spl_token::id(), false),
+        ],
+        data: PermissionedTokenInstruction::InitializeMint { decimals }.try_to_vec()?,
+    })
+}
+
+pub fn create_initialize_account_instruction(
+    mint: &Pubkey,
+    owner: &Pubkey,
+    payer: &Pubkey,
+    upstream_authority: &Pubkey,
+) -> Result<Instruction, ProgramError> {
+    let account = get_associated_token_address(owner, mint);
+    let (freeze_authority, _) =
+        Pubkey::find_program_address(&[upstream_authority.as_ref()], &crate::id());
+    Ok(Instruction {
+        program_id: crate::id(),
+        accounts: vec![
+            AccountMeta::new(account, false),
+            AccountMeta::new_readonly(*owner, false),
+            AccountMeta::new(*payer, true),
+            AccountMeta::new_readonly(*upstream_authority, true),
+            AccountMeta::new_readonly(freeze_authority, false),
+            AccountMeta::new_readonly(*mint, false),
+            AccountMeta::new_readonly(system_program::id(), false),
+            AccountMeta::new_readonly(sysvar::rent::id(), false),
+            AccountMeta::new_readonly(spl_associated_token_account::id(), false),
+            AccountMeta::new_readonly(spl_token::id(), false),
+        ],
+        data: PermissionedTokenInstruction::InitializeAccount.try_to_vec()?,
+    })
+}
+
+pub fn create_mint_to_instruction(
+    mint: &Pubkey,
+    owner: &Pubkey,
+    upstream_authority: &Pubkey,
+    amount: u64,
+) -> Result<Instruction, ProgramError> {
+    let account = get_associated_token_address(owner, mint);
+    let (mint_authority, _) = Pubkey::find_program_address(&[mint.as_ref()], &crate::id());
```

Comment:
nit: it'd be better to refactor this into a separate function, and to put in a prefix seed like `b"mint-authority"` to totally separate the domain from the upstream authority